lib/Backends/OpenCL/OpenCL.cpp

```diff
@@ -542,32 +544,35 @@ static void topK(Tensor &outW, Tensor &indW, Tensor &inW, size_t k) {
   }
 }
 
-llvm::Error OpenCLFunction::execute(ExecutionContext *context) {
-  (void)context;
-
-  auto clBindings = static_cast<runtime::OpenCLDeviceBindings *>(
-      context->getDeviceBindings());
-
-  deviceBuffer_ = clBindings->deviceBuffer;
-  deviceId_ = clBindings->deviceId;
-  commands_ = clBindings->commandQueue;
-  context_ = clBindings->context;
+llvm::Error OpenCLFunction::createKernels(cl_command_queue queue,
+                                          cl_device_id deviceId,
+                                          cl_mem deviceBuffer) {
+  deviceBuffer_ = deviceBuffer;
 
-  kernelProfiling_ = clDoProfile || getTraceInfo().autoInstrumented;
+  std::unordered_map<Value const *, size_t> instrToIdx;
```

Comment:
Also I've got a question here: Do you assume that you always launch one kernel per instruction? If you do it could be wrong. We could in theory launch multiple kernels for a single instruction.